python/singa/sonnx.py

```diff
@@ -18,272 +18,359 @@
+    def batchnorm(inputs, attrs):
+        x = inputs[0]
+        factor = attrs['momentum']
+        if x.device.id() == -1:
+            raise NotImplementedError
+        else:
+            handle = singa.CudnnBatchNormHandle(factor, x.data)
+            handle.device_id = inputs[0].device.id()
+        return handle
+
+UnaryOp = {'Relu': autograd.relu,
+           'Softmax': autograd.softmax,
+           'Flatten': autograd.flatten,
+           'Tanh': autograd.tanh,
+           'Sigmoid': autograd.sigmoid}
+BinaryOp = {'Add': autograd.add,
+            'Mul': autograd.mul,
+            'MatMul': autograd.matmul}
+
+OtherOp = {'Conv': (Handle.conv, autograd.conv2d),
+           'MaxPool': (Handle.max_pool, autograd.pooling_2d),
+           'AveragePool': (Handle.avg_pool, autograd.pooling_2d),
+           'BatchNormalization': (Handle.batchnorm, autograd.batchnorm_2d)
+           }
+
+
+class SingaBackendRep(backendRep):
+
+    def __init__(self, model, device, tensor_dict):
         '''
-        input model
-
-        return: model and model dictionary
+        Args:
+            model: onnx model proto
+            device: singa device
+            tensor_dict: dict for weight tensors
         '''
-
-        modeldic = {}
-        for i in model.graph.node:
-            if (i.op_type == 'Constant'):
-                modeldic[str(i.output[0])] = tensor.Tensor(device=device,data=numpy_helper.to_array(i.attribute[0].t),requires_grad=True, stores_grad=True)
-
-        return model,modeldic
-
+        self.model = model
+        self.device = device
+        self.tensor_dict = tensor_dict
+        self.handle_dict = {}
 
     @staticmethod
-    def find_name(model,name):
-        for i in model.graph.node:
-            if (i.name == name):
-                return i
-
-
-    @staticmethod
-    def find_shape(model,input):
+    def run_node(node, tensors, handles):
         '''
-        # find weight shape for layers
+        Args:
+            node: onnx node proto
+            tensors: dict from tensor name to tensor
+            handles: dict from node name to handle
+        '''
+        inputs = [tensors[x] for x in node.input]
+        outputs = node.output
+        attrs = attribute2dict(node)
+        op = node.op_type
+        if op in UnaryOp:
+            tensors[outputs[0]] = UnaryOp[op](inputs[0])
+        elif op in BinaryOp:
+            tensors[outputs[0]] = BinaryOp[op](inputs[0], inputs[1])
+        elif op in OtherOp:
+            handle, forward = OtherOp[op]
+            if node.name not in handles:
+                handles[node.name] = handle(inputs, attrs)
+            tensors[outputs[0]] = forward(handles[node.name], *inputs)
+        elif op == 'Concat':
+            tensors[outputs[0]] = autograd.cat(tuple(inputs), attrs['axis'])
+        else:
+            raise NotImplementedError('Not supported op: {}'.format(op))
+
+    def run(self, input):
+        # input_dict: dict from input name to numpy array
```

Comment:
input has singa tensor or numpy array?

Reply:
if singa tensor, then input can be a list of tensors (each tensor's name field must be initialized).

Reply:
singa tensor